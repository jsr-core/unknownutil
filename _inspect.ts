const defaultThreshold = 20;

export type InspectOptions = {
  // The maximum number of characters of a single attribute
  threshold?: number;
};

/**
 * Inspect a value
 */
export function inspect(value: unknown, options: InspectOptions = {}): string {
  if (value === null) {
    return "null";
  } else if (Array.isArray(value)) {
    return inspectArray(value, options);
  }
  switch (typeof value) {
    case "string":
      return JSON.stringify(value);
    case "bigint":
      return `${value}n`;
    case "object":
      if (value.constructor?.name !== "Object") {
        return value.constructor?.name;
      }
      return inspectRecord(value as Record<PropertyKey, unknown>, options);
    case "function":
      return value.name || "(anonymous)";
  }
  return value?.toString() ?? "undefined";
}

function inspectArray(value: unknown[], options: InspectOptions): string {
  const { threshold = defaultThreshold } = options;
  const vs = value.map((v) => inspect(v, options));
  const s = vs.join(", ");
  if (s.length <= threshold) return `[${s}]`;
  const m = vs.join(",\n");
  return `[\n${indent(2, m)}\n]`;
}

function inspectRecord(
  value: Record<PropertyKey, unknown>,
  options: InspectOptions,
): string {
  const { threshold = defaultThreshold } = options;
  const vs = [...Object.keys(value), ...Object.getOwnPropertySymbols(value)]
    .map((k) => `${k.toString()}: ${inspect(value[k], options)}`);
  const s = vs.join(", ");
  if (s.length <= threshold) return `{${s}}`;
  const m = vs.join(",\n");
  return `{\n${indent(2, m)}\n}`;
}

function indent(level: number, text: string): string {
  const prefix = " ".repeat(level);
  return text.split("\n").map((line) => `${prefix}${line}`).join("\n");
}
