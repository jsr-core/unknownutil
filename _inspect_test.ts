import { assertSnapshot } from "@std/testing/snapshot";
import { inspect } from "./_inspect.ts";

Deno.test("inspect", async (t) => {
  await t.step("string", async (t) => {
    await assertSnapshot(t, inspect("hello world"));
  });
  await t.step("number", async (t) => {
    await assertSnapshot(t, inspect(100));
  });
  await t.step("bigint", async (t) => {
    await assertSnapshot(t, inspect(100n));
  });
  await t.step("boolean", async (t) => {
    await assertSnapshot(t, inspect(true));
  });
  await t.step("array", async (t) => {
    await assertSnapshot(t, inspect([]));
    await assertSnapshot(t, inspect([0, 1, 2]));
    await assertSnapshot(t, inspect([0, "a", true]));
    await assertSnapshot(t, inspect([0, [1, [2]]]));
  });
  await t.step("record", async (t) => {
    await assertSnapshot(t, inspect({}));
    await assertSnapshot(t, inspect({ a: 0, b: 1, c: 2 }));
    await assertSnapshot(t, inspect({ a: "a", b: 1, c: true }));
    await assertSnapshot(t, inspect({ a: { b: { c: 0 } } }));
    await assertSnapshot(t, inspect({ [Symbol("a")]: 0 }));
    await assertSnapshot(t, inspect({ a: 0, [Symbol("b")]: 1, c: true }));
    await assertSnapshot(
      t,
      inspect({ [Symbol("a")]: { [Symbol("b")]: { [Symbol("c")]: 0 } } }),
    );
  });
  await t.step("function", async (t) => {
    await assertSnapshot(t, inspect(inspect));
    await assertSnapshot(t, inspect(() => {}));
  });
  await t.step("null", async (t) => {
    await assertSnapshot(t, inspect(null));
  });
  await t.step("undefined", async (t) => {
    await assertSnapshot(t, inspect(undefined));
  });
  await t.step("symbol", async (t) => {
    await assertSnapshot(t, inspect(Symbol("a")));
  });
  await t.step("date", async (t) => {
    await assertSnapshot(t, inspect(new Date()));
  });
  await t.step("promise", async (t) => {
    await assertSnapshot(t, inspect(new Promise(() => {})));
  });
});
