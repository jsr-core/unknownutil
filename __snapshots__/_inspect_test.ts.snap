export const snapshot = {};

snapshot[`inspect > string 1`] = `'"hello world"'`;

snapshot[`inspect > number 1`] = `"100"`;

snapshot[`inspect > bigint 1`] = `"100n"`;

snapshot[`inspect > boolean 1`] = `"true"`;

snapshot[`inspect > array 1`] = `"[]"`;

snapshot[`inspect > array 2`] = `"[0, 1, 2]"`;

snapshot[`inspect > array 3`] = `'[0, "a", true]'`;

snapshot[`inspect > array 4`] = `"[0, [1, [2]]]"`;

snapshot[`inspect > record 1`] = `"{}"`;

snapshot[`inspect > record 2`] = `"{a: 0, b: 1, c: 2}"`;

snapshot[`inspect > record 3`] = `
'{
  a: "a",
  b: 1,
  c: true
}'
`;

snapshot[`inspect > record 4`] = `"{a: {b: {c: 0}}}"`;

snapshot[`inspect > record 5`] = `"{Symbol(a): 0}"`;

snapshot[`inspect > record 6`] = `
"{
  a: 0,
  c: true,
  Symbol(b): 1
}"
`;

snapshot[`inspect > record 7`] = `
"{
  Symbol(a): {
    Symbol(b): {Symbol(c): 0}
  }
}"
`;

snapshot[`inspect > function 1`] = `"inspect"`;

snapshot[`inspect > function 2`] = `"(anonymous)"`;

snapshot[`inspect > null 1`] = `"null"`;

snapshot[`inspect > undefined 1`] = `"undefined"`;

snapshot[`inspect > symbol 1`] = `"Symbol(a)"`;

snapshot[`inspect > date 1`] = `"Date"`;

snapshot[`inspect > promise 1`] = `"Promise"`;
